store/src/java/com/zimbra/cs/imap/ImapSearch.java

```diff
@@ -35,7 +35,7 @@
 
     static boolean isAllMessages(ImapFolder i4folder, Set<ImapMessage> i4set) {
         int size = i4set.size() - (i4set.contains(null) ? 1 : 0);
-        return size == i4folder.getSize();
+        return size > ImapFolder.LARGEST_FOLDER_BATCH || size == i4folder.getSize();
```

Comment:
This goes back to what I said about Travis's original proposed fix.  My assumption would be that `isAllMessages` means `i4set` contains all messages.  This change makes it "contains all messages or is larger than the batch size".  With this change, certainly need to rename the method to something more accurate.  My worry is that it actually introduces a bug.  See comments against unit test on how a SOAP IMAP test might prove one way or the other.

Reply:
@fatkudu you are correct. This change modifies how many records are returned by UID SEARCH with a range that covers a subset of messages. Remote IMAP returns 791 messages for 'UID SEARCH 10:800 UNDELETED' before this change and 1000 messages after this change regardless of whether the selected folder is virtual or regular.